modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/checkpoint/CheckpointHistory.java

```diff
@@ -400,15 +397,17 @@ private boolean removeCheckpoint(CheckpointEntry checkpoint) {
             return false;
         }
 
-        synchronized (earliestCp) {
+        synchronized (earliestCpGrps) {
             CheckpointEntry deletedCpEntry = histMap.remove(checkpoint.timestamp());
 
-            CheckpointEntry oldestCpInHistory = firstCheckpoint();
+            CheckpointEntry oldestCpInHist = firstCheckpoint();
 
-            for (Map.Entry<GroupPartitionId, CheckpointEntry> grpPartPerCp : earliestCp.entrySet()) {
-                if (grpPartPerCp.getValue() == deletedCpEntry)
-                    grpPartPerCp.setValue(oldestCpInHistory);
-            }
+            earliestCpGrps.stream()
+                .map(cacheGrpCtxSupplier::get)
+                .filter(Objects::nonNull)
+                .flatMap(grpCtx -> grpCtx.topology().localPartitions().stream())
```

Comment:
Here, you do not handle nor index nor partition without history (0 earliest checkpoint timestamp).

Reply:
Below we have such a check, which, in theory, also covers your proposal.
`.filter(localPart -> localPart.earliestCpTs() == deletedCpEntry.timestamp())`